relay-cardinality/src/redis.rs

```diff
@@ -1,27 +1,131 @@
-use std::{collections::HashSet, time::SystemTime};
+use std::time::SystemTime;
 
 use itertools::Itertools;
 use relay_redis::{
-    redis::{self, ToRedisArgs},
+    redis::{self, FromRedisValue, Script},
     Connection, RedisPool,
 };
 use relay_statsd::metric;
 
 use crate::{
     limiter::{CardinalityScope, Entry, EntryId, Limiter, Rejection},
-    statsd::CardinalityLimiterCounters,
+    statsd::{CardinalityLimiterCounters, CardinalityLimiterHistograms},
     OrganizationId, Result,
 };
 
 /// Key prefix used for Redis keys.
 const KEY_PREFIX: &str = "relay:cardinality";
-/// Maximum amount of arguments for a Redis `SADD` call.
-const MAX_SADD_ARGS: usize = 1000;
+
+struct CardinalityScript(Script);
+
+/// Status wether an entry/bucket is accepted or rejected by the cardinality limiter.
+#[derive(Clone, Copy, Debug)]
+enum Status {
+    /// Item is rejected.
+    Rejected,
+    /// Item is accepted.
+    Accepted,
+}
+
+impl Status {
+    fn is_rejected(&self) -> bool {
+        matches!(self, Self::Rejected)
+    }
+}
+
+impl FromRedisValue for Status {
+    fn from_redis_value(v: &redis::Value) -> redis::RedisResult<Self> {
+        let accepted = bool::from_redis_value(v)?;
+        Ok(if accepted {
+            Self::Accepted
+        } else {
+            Self::Rejected
+        })
+    }
+}
+
+#[derive(Debug)]
+struct CardinalityScriptResult {
+    cardinality: u64,
+    rejected: usize,
+    bitmap: Vec<Status>,
+}
+
+impl FromRedisValue for CardinalityScriptResult {
+    fn from_redis_value(v: &redis::Value) -> redis::RedisResult<Self> {
+        let Some(seq) = v.as_sequence() else {
+            return Err(redis::RedisError::from((
+                redis::ErrorKind::TypeError,
+                "Expected a sequence from the cardinality script",
+            )));
+        };
+
+        let mut iter = seq.iter();
+
+        let cardinality = iter
+            .next()
+            .ok_or_else(|| {
+                redis::RedisError::from((
+                    redis::ErrorKind::TypeError,
+                    "Expected cardinality as the first result from the cardinality script",
+                ))
+            })
+            .and_then(FromRedisValue::from_redis_value)?;
+
+        let mut bitmap = Vec::with_capacity(iter.len() - 1);
```

Comment:
Does `iter.len()` return the original length, even if the iterator already consumed one element?

Reply:
Thanks, that is a refactoring but and was supposed to use the underlaying source not the iterator!